node-api: get Node API version used by addon

vmoroz opened the pull request:
This is to address the issue #45657.

The main goal of the PR is to establish a new mechanism to implement Node-API changes that affect not only the API shape, but also its behavior.
We are going to associate Node-API functions and new Node-API behavior with new Node-API version.
Node-API modules can choose which version of Node-API they target.
By default, they target the current Node-API version 8.
To target newer versions of Node-API they must explicitly opt-in by defining `napi_module_get_api_version_v1` export function that returns targeted version.

In this PR, the `NAPI_MODULE` and `NAPI_MODULE_INITIALIZER` macros are changed to define the `napi_module_get_api_version_v1` function that reports the value of the `NAPI_VERSION` macro. C++ developers can control the targeted Node-API version by defining the `NAPI_VERSION` macro as they did before. The key difference is that now this macro not only affects the set of functions that module uses but also affects the Node-API implementation behavior.

Node.js is going to validate the version requested by the addon.
- If the version is below version 8, then it will be internally switched to version 8 because we did not implement the behavior change before version 8.
- If the version is above the `NAPI_VERSION` the Node.js compiled for and it is not `NAPI_VERSION_EXPERIMENTAL`, then Node.js will throw an error on module initialization.

Note that modules targeting `NAPI_VERSION_EXPERIMENTAL` are compatible with any Node-API version the Node.js was compiled with. Node-API in that case will use the latest functions and behaviors. Such module should call `napi_get_version` function to get the latest stable Node-API version offered by current Node.js instance. Then it must adapt for the different behavior based on the version.

In this PR, we change the behavior of the `napi_create_reference` function for Node-API versions higher than 8. The new behavior allows the creation of an `napi_ref` for any `napi_valuetype` instead of only `napi_object`, `napi_function`, `napi_external`, and `napi_symbol`. One of the main goals is to enable the use of `napi_ref` for `napi_string`.

The `napi_ref` for the new primitive types immediately releases them when the ref count becomes 0 because they cannot be weak references. For the previously supported object and symbol types, the behavior did not change: these values become weak references and can be collected at any time. Note that the globally registered symbols created with the `Symbol.for()` function call and well-known symbols such as `Symbol.iterator` are always kept alive because they are never collected by the garbage collector.

KevinEady commented:
We discussed this draft's approach in the 9 Dec 2022 Node-API meeting. One question that was brought up is discussion about using two functions for the register-by-symbol method (one for getting version and one for initializing the module) and if this is a good approach versus just using a single method. It was determined that using two methods would be more beneficial because we need to know the targeted runtime version _before_ the module initialization code runs, since this code can run arbitrary JavaScript.

legendecas commented on src/node_api.cc:
```diff
@@ -597,13 +599,16 @@ static void napi_module_register_cb(v8::Local<v8::Object> exports,
       module,
       context,
-      static_cast<const napi_module*>(priv)->nm_register_func);
+      static_cast<const napi_module*>(priv)->nm_register_func,
```
We can reduce duplication of casting here:

```suggestion
  const napi_module* m = static_cast<const napi_module*>(priv);
  napi_module_register_by_symbol(
      exports,
      module,
      context,
      m->nm_register_func,
      static_cast<int32_t>(m->nm_api_version));
```

vmoroz replied:
Good point! I applied the suggestion.

legendecas commented on src/node_api.h:
```diff
@@ -119,6 +140,73 @@ typedef struct napi_module {
+  NAPI_MODULE_XV(modname, regfunc, NAPI_VERSION, priv, flags)
+
+#define NAPI_CONCAT_HELPER(text1, text2) text1##text2
+#define NAPI_CONCAT(text1, text2) NAPI_CONCAT_HELPER(text1, text2)
```
Looks like this is the renamed `NAPI_MODULE_INITIALIZER_X`. When `NAPI_EXPERIMENTAL` is defined, `NAPI_MODULE_INITIALIZER_X` is not available anymore. Is it ok for us to remove undocumented macros?

vmoroz replied:
Usually, it must be OK to remove such macros, but we cannot say if some code did not take dependency on it. I restored it.

legendecas commented:
Thank you for working on this! I know this is still a draft but I'll continue reviewing.

gabrielschulhof commented on src/node_api.h:
```diff
@@ -54,11 +55,20 @@ typedef struct napi_module {
 #define NAPI_MODULE_INITIALIZER_BASE napi_register_module_v
 #endif
 
+#define NAPI_MODULE_GET_API_VERSION_BASE napi_module_get_api_version_v
```
Since this is a new Node-API-related preprocessor directive in a public header, can we rename it and all newly added preprocessor directives to `NODE_API_*` (`NODE_API_MODULE_GET_API_VERSION_BASE` in this case and `NODE_API_MODULE_GET_API_VERSION` below)?

vmoroz replied:
Good point! I will do it.

vmoroz replied:
Changed all `napi` prefixes to `node_api` for the new APIs.

gabrielschulhof reviewed (commented):
Mostly LGTM. Can we combine the two tests into a single test that uses two addons? So, like,
```gyp
{
  "targets": [
    {
      "target_name": "test_reference_all_types",
      "sources": [ "test_reference_by_node_api_version.c" ],
      "defines": [ "NAPI_EXPERIMENTAL" ],
    },
    {
      "target_name": "test_reference_obj_only",
      "sources": [ "test_reference_by_node_api_version.c" ],
      "defines": [ "NAPI_EXPERIMENTAL" ],
    }
  ]
}
```

... and then do some `#ifdef`-ing in the source file? The test JS can then `require()` each add-on and perform a fixed set of tests with two different sets of expectations as passed-in parameters.

vmoroz commented:
> Mostly LGTM. Can we combine the two tests into a single test that uses two addons? So, like,

Let me see if it can be done.

gabrielschulhof commented:
> > Mostly LGTM. Can we combine the two tests into a single test that uses two addons? So, like,
> 
> Let me see if it can be done.

It may also be possible to distill common bits into one source file which is listed in the sources section of both targets and then have two source files, each of which has the version-specific changes, one for each target. That should reduce the need for `#ifdef`s.

vmoroz commented:
> > > Mostly LGTM. Can we combine the two tests into a single test that uses two addons? So, like,
> > 
> > 
> > Let me see if it can be done.
> 
> It may also be possible to distill common bits into one source file which is listed in the sources section of both targets and then have two source files, each of which has the version-specific changes, one for each target. That should reduce the need for `#ifdef`s.

I combined both tests and found that the C code between them was exactly the same. So, there was no need for #ifdefs. I also noticed that the JS code was mostly the same and I combined them too with minimal changes. Now the test looks pretty cool because it shows how we can create two different addons from the same C code that have different behavior depending on the targeted Node-API version. Then we use these addons in the same JS code.

gabrielschulhof commented on test/node-api/test_reference_by_node_api_version/test.js:
```diff
@@ -0,0 +1,108 @@
+// Testing API calls for Node-API references.
+// We compare their behavior between Node-API version 8 and later.
+// In version 8 references can be created only for object, function,
+// and symbol types. While in newer versions they can be created for
```
```suggestion
// and symbol types, while in newer versions they can be created for
```

vmoroz replied:
fixed

vmoroz commented on doc/api/n-api.md:
```diff
@@ -1787,18 +1794,14 @@ NAPI_EXTERN napi_status napi_get_reference_value(napi_env env,
-* `[in] ref`: `napi_ref` for which we requesting the corresponding `Object`.
-* `[out] result`: The `napi_value` for the `Object` referenced by the
-  `napi_ref`.
+* `[in] ref`: `napi_ref` for which we requesting the corresponding value.
```
fixed

vmoroz commented on src/node_binding.cc:
```diff
@@ -484,7 +490,12 @@ void DLOpen(const FunctionCallbackInfo<Value>& args) {
       } else if (auto napi_callback = GetNapiInitializerCallback(dlib)) {
-        napi_module_register_by_symbol(exports, module, context, napi_callback);
+        int32_t module_api_version = 0;
+        if (auto get_version = GetNapiAddonGetApiVersionCallback(dlib)) {
```
We can, but I believe it is a better coding style to keep variable declaration closer to its use. It reduces the lifespan of the variable and the risk of introducing bugs. It is also easier to read.

vmoroz commented:
@legendecas , @mhdawson . the PR is ready to be merged. Could you please take a look at it?

legendecas commented on src/api/environment.cc:
```diff
@@ -885,7 +885,8 @@ void AddLinkedBinding(Environment* env,
             context,
-            reinterpret_cast<napi_addon_register_func>(priv));
+            reinterpret_cast<napi_addon_register_func>(priv),
+            NODE_API_DEFAULT_MODULE_API_VERSION);
```
Can we add an optional parameter so that the target version can be set with `AddLinkedBinding` since registering with `napi_addon_register_func` is the preferred approach to register a linked node-api addon?

vmoroz replied:
Good suggestion! The code is changed. It is now a bit more complex than before, but it works. There are new unit tests to verify it.

legendecas commented on test/node-api/test_reference_by_node_api_version/test_reference_by_node_api_version.c:
```diff
@@ -0,0 +1,175 @@
+#define NAPI_EXPERIMENTAL
```
The `NAPI_VERSION` should have been defined in the binding.gyp.

```suggestion
```

vmoroz replied:
Good catch! It was a bug after I merged two tests. Fixed.

legendecas commented on src/js_native_api_v8.h:
```diff
@@ -51,8 +51,13 @@ class Finalizer;
+                      int32_t module_api_version)
+      : isolate(context->GetIsolate()),
+        context_persistent(isolate, context),
+        module_api_version(module_api_version != 0
```
IIUC, if the callsite of the callback `GetNapiAddonGetApiVersionCallback` set the `module_api_version` defaults to `NODE_API_DEFAULT_MODULE_API_VERSION` instead of `0`, this conditional branch can be removed.

vmoroz replied:
I would prefer doing such checks closer to the use of the variable because we cannot know all possible callers in future. 
Anyway, I followed the suggestion and removed the check.

legendecas reviewed (commented):
Overall LGTM % nits. Thank you for working on this!

vmoroz commented:
It seems that the test-asan failures are valid - they are about the memory leaks in the new unit tests for `AddLinkedBinding` function where I create `napi_ref`. I am going to fix it.

gabrielschulhof commented on src/node_api.cc:
```diff
@@ -152,10 +153,22 @@ class BufferFinalizer : private Finalizer {
+             module_api_version != NAPI_VERSION_EXPERIMENTAL) {
+    node::Environment* node_env = node::Environment::GetCurrent(context);
+    CHECK_NOT_NULL(node_env);
+    node_env->ThrowError("Node API module targets unsupported version.");
```
Can we make this more verbose?

```suggestion
    node_env->ThrowError(
        module_filename + "requires Node-API version " +
        std::to_string(module_api_version) + ", but this version of Node.js "
        "only supports version " #NAPI_VERSION " add-ons.");
```

vmoroz replied:
Good suggestion! I applied your code and extracted it into a separate method since we need it in two different places.

gabrielschulhof commented on src/node_api.cc:
```diff
@@ -623,10 +636,42 @@ static void napi_module_register_cb(v8::Local<v8::Object> exports,
+// Currently AddLinkedBinding is the only user of this function.
+node::addon_context_register_func get_node_api_context_register_func(
+    node::Environment* node_env, int32_t module_api_version) {
+  if (module_api_version <= NODE_API_DEFAULT_MODULE_API_VERSION) {
```
Can we add a `static_assert` here so that, if we increment `NAPI_VERSION`, the build will fail if we do not remember to augment this function? The message of the `static_assert` should remind the reader to also update the conditions under which the `static_assert` will fail, because we will likely need to explicitly write the expected `NAPI_VERSION` as a numeric literal. Re https://github.com/nodejs/node/issues/47664.

vmoroz replied:
Great idea! Added the `static_assert`.

gabrielschulhof commented on test/node-api/test_reference_by_node_api_version/binding.gyp:
```diff
@@ -0,0 +1,14 @@
+      "defines": [ "NAPI_VERSION=8" ],
+    }
+  ]
+}
```
nit: could use a trailing newline here.

vmoroz replied:
Fixed

vmoroz commented:
@mhdawson, @legendecas, @gabrielschulhof, @addaleax, the PR is passing all checks and I have addressed all the feedback so far. Could you please have another look?

toyobayashi commented:
https://github.com/nodejs/node/blob/e85db63abd1fd69c8339dd78bf90c1773a2bd82a/src/js_native_api_v8.cc#L555

> Previously, when the ref count becomes 0, the napi_ref holding napi_symbol continued to keep it. In this PR, we fix it

Found https://github.com/nodejs/node/issues/39131 and https://github.com/nodejs/node/pull/39926, why `Symbol` is not allowed to be weak any more? [Symbols as WeakMap keys](https://github.com/tc39/proposal-symbols-as-weakmap-keys) is available on latest Chrome and Node.js v20.

toyobayashi commented:
Thanks for your explanation, I had thought `SetWeak` worked for `Symbol` before. Relying on the Symbol as WeakMap keys proposal, emnapi has already implemented weak symbol references, so I was confused about the changes here before comment. I will sync the behavior of this PR into emnapi :)

vmoroz commented:
> Thanks for your explanation, I had thought `SetWeak` worked for `Symbol` before. Relying on the Symbol as WeakMap keys proposal, emnapi has already implemented weak symbol references, so I was confused about the changes here before comment. I will sync the behavior of this PR into emnapi :)

@toyobayashi, thank you for bringing this issue! I had found that the situation with symbols is much more complex than I thought before. Only local non-registered symbols created with the `Symbol` function can have the weak reference semantic. Globally registered symbols created with the `Symbol.for` method cannot be weak references. The [MDN WeakRef doc](https://developer.mozilla.org/en-US/docs/web/javascript/reference/global_objects/weakref) talks about. Since I only used registered global symbols, I could not see them becoming weak.

The code is updated to differentiate between the two types based on the result of [`Symbol.keyFor`]() function. It returns `undefined` for local symbols. Unfortunately, this method is not perfect. It does not work for the well-known predefined symbols like `Symbol.iterator`. I do not see a simple way to address the well-known symbols yet. It would be great if V8 exposes this info in the public API.

legendecas commented on src/js_native_api_v8.cc:
```diff
@@ -457,6 +457,49 @@ inline napi_status Wrap(napi_env env,
+    v8::Local<v8::Context> context = env->context();
+    v8::Local<v8::Function> symbol_key_for;
+    if (env->symbol_key_for_.IsEmpty()) {
+      v8::Local<v8::Object> global = context->Global();
```
We should get `Symbol.keyFor` reference from the per-context primordial instead. See https://github.com/nodejs/node/blob/main/doc/contributing/primordials.md for more details. The per-context primordial is available from C++ with `env->primordials()`.

vmoroz replied:
I removed this code per your recommendation.

legendecas reviewed (commented):
The Node-API version part LGTM.

However, I think the symbol weakness doesn't align with the WeakMap keys or the target of WeakRefs semantics as the well-known symbols are allowed in these conditions. What do you think about splitting the `napi_ref` change to a new PR so that we can land the runtime version first?

legendecas commented:
Calling into JS when creating a `napi_ref` should be avoided. Can we document that it is undefined behavior when creating a weak reference with `napi_ref` on primitive values instead? i.e. implementation can release the values immediately or never release the values.

vmoroz commented:
> However, the symbol weakness doesn't align with the WeakMap keys or the target of WeakRefs semantics as the well-known symbols are allowed in these conditions. What do you think about splitting the `napi_ref` change to a new PR so that we can land the runtime version first?

No, I do not want to split up the `napi_ref` into a different PR from the version. It is the main functionality I must add to Node-API to support references to all value types required by JSI. I would rather drop the versioning support from this PR.
Also, there were discussions before saying that it does not make sense to add versioning support if we have nothing to put into the new version. The `napi_ref` change is quite minimal in that regard.

In the latest iteration the functionality for object and symbol `napi_ref` is unchanged. We only add support for the remaining value types. Thus, in regard to `WeakMap` keys or `WeakRefs` there are no changes comparing to what we have today.

gabrielschulhof reviewed (commented):
I got partway through a review. Will look further.

gabrielschulhof commented on test/node-api/test_reference_by_node_api_version/test.js:
```diff
@@ -0,0 +1,124 @@
+    const booleanValue = false;
+    const numberValue = 42;
+    const stringValue = 'test_string';
+    const globalSymbolValue = Symbol.for('test_symbol_global');
```
Do we still need this distinction between global and local symbols?

vmoroz replied:
Yes, we need it because they behave differently when the `napi_ref` ref count becomes 0 and we run GC.
My previous test did not make a distinction between the two types because I only used a global symbol.
It caused me wrongly to believe that Symbols are never weak. 
It is good to show/test the different behavior for these two types of symbols.

vmoroz commented on doc/api/n-api.md:
```diff
@@ -1668,21 +1679,24 @@ will return `NULL` for the returned `napi_value`. An attempt to call
 [`napi_reference_ref`][] for a reference whose object has been collected
 results in an error.
 
+In Node-API version 8 and before we can only create references for a limited
```
Fixed. I asked ChatGPT to create readable text for me, but it was not aware about the typical technical docs' limitations. :)

vmoroz commented on doc/api/n-api.md:
```diff
@@ -1700,15 +1714,18 @@ NAPI_EXTERN napi_status napi_create_reference(napi_env env,
 * `[in] env`: The environment that the API is invoked under.
-* `[in] value`: `napi_value` representing the `Object` to which we want a
-  reference.
+* `[in] value`: `napi_value` to which we want to create a reference.
```
Updated. I went through the whole file and fixed other occurrences of "we".

mhdawson commented on src/js_native_api_v8.cc:
```diff
@@ -457,6 +457,18 @@ inline napi_status Wrap(napi_env env,
+// method cannot be weak references because they are never collected.
+//
+// Currently, V8 has no API to detect if a symbol is local or global.
+// Until we have a V8 API for it, we consider that all symbols can be weak.
```
This does match current behaviour but also means that we'll have a leak right?  Since we can change behaviour in a new Node-API version without affecting existing applications maybe the right thing to do is to consider all symbols as not being week so that the leak does not occur?

mhdawson replied:
@vmoroz thanks for the clarification. I was thinking about the auto-delete we have in RefBase but looking more closely I see that won't be related to direct use of references.

mhdawson commented:
@vmoroz just one question, otherwise all looks good to me.  Will wait to add approval until we agree on the answer to that question.

mhdawson reviewed (approved):
LGTM

legendecas commented:
Sorry for going in circles about the PR. Thank you for pushing this forward!

mhdawson commented:
@vmoroz, thanks for all of the hard on this, landed in c542d3a1d30d

Eladrian05 reviewed (commented):
:)
